### PR TITLE
Transform `if` / `else if` into flat `if` / `elseif` when possible

### PR DESCRIPTION
**Pull request by herndlm**
Closes https://github.com/phpstan/phpstan/issues/10113

Alternative to https://github.com/phpstan/phpstan-src/pull/2728

### CODE REVIEW

**Comment by herndlm** [comment]
Looking good, but apparently I made Rector on Windows angry somehow?

**Comment by herndlm** [comment]
any idea what the (rector) issue might be here? it kept happening after a rebase, I don't want to rebase unnecessarily. wonder if I need to bump some rector cache key?

File: src/Parser/RichParser.php [hunk 1]
```diff
@@ -66,6 +66,9 @@ public function parseString(string $sourceCode): array
 		$traitCollectingVisitor = new TraitCollectingVisitor();
 		$nodeTraverser->addVisitor($traitCollectingVisitor);
 
+		$transformElseIfToFlatElseIfVisitor = new TransformElseIfToFlatElseIfVisitor();
+		$nodeTraverser->addVisitor($transformElseIfToFlatElseIfVisitor);
```
Comment (ondrejmirtes) [inline src/Parser/RichParser.php]: I'm a bit surprised this works. I'm not sure about the order and effects the visitors are called in. I'd expect the other visitors not being able to see effects from this one.

I'd be a little bit calmer if the AST was transformed with `TransformElseIfToFlatElseIfVisitor` by a separate NodeTraverser and only after that it was passed to the current `$nodeTraverser`.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
Also: Please write a custom rule in `tests` that just verifies the AST is transformed as expected.

So it should hook onto `If_` and just report how many `elseifs` it sees.

The Rector issue might or might not be fixed by other suggestion :)

**Comment by herndlm** [comment]
Got inspired by `CleaningParserTest` and ended up writing the test a bit different. WDYT? the (stolen) core part could be useful for more traverse/transform tests if needed, but it felt not right to try to make it generic now.
Rector still doesn't like me here apparently :/

**Comment by herndlm** [comment]
I believe this one is ready (for review) again. I was just waiting for you to switch to the simplified downgrader :)

**Comment by ondrejmirtes** [comment]
The failures in Rector worry me a bit: https://github.com/phpstan/phpstan-src/actions/runs/7420725636/job/20192723240

**Comment by herndlm** [comment]
I don't know Rector well enough but I guess the errors related to that if/else rector mean that the change here is not possible. Is https://github.com/phpstan/phpstan-src/pull/2728 a better alternative after all, maybe? Or something else? 🤔

**Pull request closed without merging**